manim/utils/qhull.py

```diff
@@ -0,0 +1,163 @@
+        )
+
+    def compute_normal(self, internal):
+        centered = self.coordinates - self.center
+        _, _, vh = np.linalg.svd(centered)
+        normal = vh[-1, :]
+        normal /= np.linalg.norm(normal)
+
+        if np.dot(normal, self.center - internal) < 0:
+            normal *= -1
+
+        return normal
+
+    def __hash__(self):
+        return hash(self.subfacets)
+
+    def __eq__(self, other):
+        return self.subfacets == other.subfacets
+
+
+class Horizon:
+    def __init__(self):
+        self.facets = set()
+        self.boundary = []
+
+
+class QuickHull:
+    def __init__(self, tolerance=1e-5):
+        self.facets = []
+        self.removed = set()
+        self.outside = {}
+        self.neighbors = {}
+        self.unclaimed = None
+        self.internal = None
+        self.tolerance = tolerance
+
+    def initialize(self, points):
+        # Sample Points
+        simplex = points[
+            np.random.choice(points.shape[0], points.shape[1] + 1, replace=False)
+        ]
+        self.unclaimed = points
+        self.internal = np.mean(simplex, axis=0)
+
+        # Build Simplex
+        for c in range(simplex.shape[0]):
+            facet = Facet(np.delete(simplex, c, axis=0), internal=self.internal)
+            self.classify(facet)
+            self.facets.append(facet)
+
+        # Attach Neighbors
+        for f in self.facets:
+            for sf in f.subfacets:
+                self.neighbors.setdefault(sf, set()).add(f)
+
+    def classify(self, facet):
+        if not self.unclaimed.size:
+            self.outside[facet] = (None, None)
+            return
+
+        # Compute Projections
+        projections = (self.unclaimed - facet.center) @ facet.normal
+        arg = np.argmax(projections)
+        mask = projections > self.tolerance
+
+        # Identify Eye and Outside Set
+        eye = self.unclaimed[arg] if projections[arg] > self.tolerance else None
+        outside = self.unclaimed[mask]
+        self.outside[facet] = (outside, eye)
+        self.unclaimed = self.unclaimed[~mask]
+
+    def compute_horizon(self, eye, start_facet):
+        horizon = Horizon()
+        self._recursive_horizon(eye, start_facet, horizon)
+        return horizon
+
+    def _recursive_horizon(self, eye, facet, horizon):
+        # If the eye is visible from the facet...
+        if np.dot(facet.normal, eye - facet.center) > 0:
+            # Label the facet as visible and cross each edge
+            horizon.facets.add(facet)
+            for subfacet in facet.subfacets:
+                neighbor = (self.neighbors[subfacet] - {facet}).pop()
+                # If the neighbor is not visible, then the edge shared must be on the boundary
+                if neighbor not in horizon.facets and not self._recursive_horizon(
+                    eye, neighbor, horizon
+                ):
+                    horizon.boundary.append(subfacet)
+            return 1
+        else:
+            return 0
```

Comment:
It seems CodeQL didn't like the `if ... return 1 else return 0`, it didn't realize that all the possible return cases were already covered and there were no implicit (fall through) returns.

One option would be to remove that else, so that instead of
```py
        if condition:
            # large chunk of code
            return 1
        else:
            return 0
```
it was straight up
```py
        if condition:
            # large chunk of code
            return 1
        return 0
```
However it doesn't look very nice to have that small dangling `return 0` after that large chunk of code. It would be better to do something like this, which handles the small case and returns from it immediately if necessary, and removes one indentation level from the large chunk of code:

```py
        if not condition:
            return 0
        
        # large chunk of code
        return 1
```

Moreover, those returned ints (1 and 0) are, according to the code, meant to be used as `True` and `False`, so it's better to stick to that instead:


```py
        if not condition:
            return False
        
        # large chunk of code
        return True
```
Thus, my suggestion is the following:
```suggestion
    def _recursive_horizon(
        self, eye: npt.NDArray[ManimFloat], facet: Facet, horizon: Horizon
    ) -> bool:
        eye_is_visible_from_facet = np.dot(facet.normal, eye - facet.center) > 0
        if not eye_is_visible_from_facet:
            return False
            
        # If the eye is visible, label the facet as visible and cross each edge
        horizon.facets.add(facet)
        for subfacet in facet.subfacets:
            neighbor = (self.neighbors[subfacet] - {facet}).pop()
            # If the neighbor is not visible, then the edge shared must be on the boundary
            if neighbor not in horizon.facets and not self._recursive_horizon(
                eye, neighbor, horizon
            ):
                horizon.boundary.append(subfacet)
        return True
```

Reply:
Ah, that makes sense! I was confused why CodeQL didn't like the original form event though all possible return cases were addressed, so I added the else statement hoping that would make it more explicit. I like your version better!